src/wallet/migrate.cpp

```diff
@@ -0,0 +1,771 @@
+    {
+        s >> unused2;
+        s >> page_number;
+        s >> item_len;
+
+        if (m_header.other_endian) {
+            page_number = internal_bswap_32(page_number);
+            item_len = internal_bswap_32(item_len);
+        }
+    }
+};
+
+/** A generic data page in the database. Contains fields common to all data pages. */
+class PageHeader
+{
+public:
+    uint64_t lsn;       // Log Sequence Number
+    uint32_t page_num;  // Current page number
+    uint32_t prev_page; // Previous page number
+    uint32_t next_page; // Next page number
+    uint16_t entries;   // Number of items on the page
+    uint16_t hf_offset; // High free byte page offset
+    uint8_t level;      // Btree page level
+    PageType type;      // Page type
+
+    static constexpr int64_t SIZE = 26; // The header is 26 bytes
+
+    uint32_t expected_page_num;
+    bool other_endian;
+
+    PageHeader(uint32_t page_num, bool other_endian) : expected_page_num(page_num), other_endian(other_endian) {}
+    PageHeader() = delete;
+
+    template<typename Stream>
+    void Unserialize(Stream& s)
+    {
+        s >> lsn;
+        s >> page_num;
+        s >> prev_page;
+        s >> next_page;
+        s >> entries;
+        s >> hf_offset;
+        s >> level;
+
+        uint8_t uint8_type;
+        s >> uint8_type;
+        type = static_cast<PageType>(uint8_type);
+
+        if (other_endian) {
+            lsn = internal_bswap_64(lsn);
+            page_num = internal_bswap_32(page_num);
+            prev_page = internal_bswap_32(prev_page);
+            next_page = internal_bswap_32(next_page);
+            entries = internal_bswap_16(entries);
+            hf_offset = internal_bswap_16(hf_offset);
+        }
+
+        if (expected_page_num != page_num) {
+            throw std::runtime_error("Page number mismatch");
+        }
+        if ((type != PageType::OVERFLOW_DATA && level < 1) || (type == PageType::OVERFLOW_DATA && level != 0)) {
+            throw std::runtime_error("Bad btree level");
+        }
+    }
+};
+
+/** A page of records in the database */
+class RecordsPage
+{
+public:
+    RecordsPage(const PageHeader& header) : m_header(header) {}
+    RecordsPage() = delete;
+
+    PageHeader m_header;
+
+    std::vector<uint16_t> indexes;
+    std::vector<std::variant<DataRecord, OverflowRecord>> records;
+
+    template<typename Stream>
+    void Unserialize(Stream& s)
+    {
+        // Current position within the page
+        int64_t pos = PageHeader::SIZE;
+
+        // Get the items
+        for (uint32_t i = 0; i < m_header.entries; ++i) {
+            // Get the index
+            uint16_t index;
+            s >> index;
+            if (m_header.other_endian) {
+                index = internal_bswap_16(index);
+            }
+            indexes.push_back(index);
+            pos += sizeof(uint16_t);
+
+            // Go to the offset from the index
+            int64_t to_jump = index - pos;
+            s.ignore(to_jump);
```

Comment:
Note that the argument to `ignore()` is unsigned (a `size_t`), might want to error out here if somehow `index < pos`.
(same for the occurence in `InternalPage`)

Reply:
Added a check